[mlir][IR][NFC] Make `replaceAllUsesWith` non-templatized

matthias-springer opened the pull request:
Turn `RewriterBase::replaceAllUsesWith` into a non-templatized implementation, so that it can be made virtual and be overridden in the `ConversionPatternRewriter` in a subsequent change.

This change is in preparation of adding dialect conversion support for `replaceAllUsesWith`.

ingomueller-net reviewed (approved):
Yeah, makes sense. I have contemplated that possibility already [when I added the templated version](https://reviews.llvm.org/D142525#4079227).

matthias-springer committed:
[mlir][IR][NFC] Make `replaceAllUsesWith` non-templatized

Turn `RewriterBase::replaceAllUsesWith` into a non-templatized implementation, so that it can be made virtual and be overridden in the `ConversionPatternRewriter` in a subsequent change.

This change is in preparation of adding dialect conversion support for `replaceAllUsesWith`.
